### PR TITLE
fix data race in history code (#405)

### PR DESCRIPTION
**Pull request by HeavyHorst**

### CODE REVIEW

File: edit/history/fuser.go [hunk 1]
```diff
@@ -16,16 +22,22 @@ func NewFuser(store Store) (*Fuser, error) {
...
 
 func (f *Fuser) AddCmd(cmd string) error {
 	seq, err := f.store.AddCmd(cmd)
 	if err != nil {
 		return err
 	}
+	f.Lock()
 	f.cmds = append(f.cmds, cmd)
 	f.seqs = append(f.seqs, seq)
+	f.Unlock()
```
Comment (xiaq) [inline edit/history/fuser.go]: Please use `defer`.

File: edit/history/fuser.go [hunk 2]
```diff
@@ -34,13 +46,21 @@ func (f *Fuser) AllCmds() ([]string, error) {
 	if err != nil {
 		return nil, err
 	}
-	return append(cmds, f.cmds...), nil
+	return append(cmds, f.SessionCmds()...), nil
```
Comment (xiaq) [inline edit/history/fuser.go]: Please use a read-lock on this function, and leave `f.cmds` as is.

File: edit/history/fuser.go [hunk 3]
```diff
@@ -34,13 +46,21 @@ func (f *Fuser) AllCmds() ([]string, error) {
...
 }
 
 func (f *Fuser) SessionCmds() []string {
+	f.RLock()
```
Comment (xiaq) [inline edit/history/fuser.go]: There is no need for this lock.

File: edit/history/fuser.go [hunk 4]
```diff
@@ -34,13 +46,21 @@ func (f *Fuser) AllCmds() ([]string, error) {
...
+	defer f.RUnlock()
 	return f.cmds
 }
 
+func (f *Fuser) getSeqs() []int {
+	f.RLock()
```
Comment (xiaq) [inline edit/history/fuser.go]: There is no need for this lock. Please call this function `Seqs`.

File: edit/history/fuser.go [hunk 5]
```diff
@@ -34,13 +46,21 @@ func (f *Fuser) AllCmds() ([]string, error) {
...
+	defer f.RUnlock()
+	return f.seqs
+}
+
 func (f *Fuser) Walker(prefix string) *Walker {
-	return NewWalker(f.store, f.storeUpper, f.cmds, f.seqs, prefix)
+	return NewWalker(f.store, f.storeUpper, f.SessionCmds(), f.getSeqs(), prefix)
```
Comment (xiaq) [inline edit/history/fuser.go]: Please install a read-lock on this function and use the fields directly.

**Comment by HeavyHorst** [comment]
Done ! :)

File: edit/history/fuser.go [hunk 6]
```diff
@@ -16,14 +22,20 @@ func NewFuser(store Store) (*Fuser, error) {
...
+		RWMutex:    &sync.RWMutex{},
+	}, nil
 }
 
 func (f *Fuser) AddCmd(cmd string) error {
 	seq, err := f.store.AddCmd(cmd)
 	if err != nil {
 		return err
 	}
+	f.Lock()
```
Comment (xiaq) [inline edit/history/fuser.go]: Please move the locking to the beginning of the function.

File: edit/history/fuser.go [hunk 7]
```diff
@@ -34,6 +46,8 @@ func (f *Fuser) AllCmds() ([]string, error) {
 	if err != nil {
 		return nil, err
 	}
+	f.RLock()
```
Comment (xiaq) [inline edit/history/fuser.go]: Ditto

**Comment by HeavyHorst** [comment]
I've moved the locking to the start of the functions.

**Pull request merged**